src/oob/idcert.rs

```diff
@@ -0,0 +1,526 @@
+        if let Some(ref aki) = self.extensions.authority_key_id {
+            if *aki != issuer.extensions.subject_key_id {
+                return Err(ValidationError)
+            }
+        }
+        else {
+            return Err(ValidationError);
+        }
+
+        Ok(())
+    }
+
+    /// Validates that the certificate is a valid CA certificate.
+    ///
+    /// Checks the parts that are common in normal and trust anchor CA
+    /// certificates.
+    fn validate_ca_basics(&self) -> Result<(), ValidationError> {
+        // 4.8.1. Basic Constraints: For a CA it must be present (RFC6487)
+        // und the “cA” flag must be set (RFC5280).
+        if self.extensions.basic_ca != Some(true) {
+            return Err(ValidationError)
+        }
+
+        Ok(())
+    }
+
+    /// Validates the certificate’s signature.
+    fn validate_signature(
+        &self,
+        issuer: &IdCert
+    ) -> Result<(), ValidationError> {
+        self.signed_data.verify_signature(issuer.public_key())
+    }
+}
+
+
+//--- AsRef
+
+impl AsRef<IdCert> for IdCert {
+    fn as_ref(&self) -> &Self {
+        self
+    }
+}
+
+//------------ Validity ------------------------------------------------------
+
+#[derive(Clone, Debug)]
+pub struct Validity {
+    not_before: Time,
+    not_after: Time,
+}
+
+impl Validity {
+    pub fn new(not_before: Time, not_after: Time) -> Self {
+        Validity { not_before, not_after }
+    }
+
+    pub fn take_from<S: Source>(
+        cons: &mut Constructed<S>
+    ) -> Result<Self, S::Err> {
+        cons.take_sequence(|cons| {
+            Ok(Validity::new(
+                Time::take_from(cons)?,
+                Time::take_from(cons)?,
+            ))
+        })
+    }
+
+    pub fn validate(&self) -> Result<(), ValidationError> {
+        self.not_before.validate_not_before()?;
+        self.not_after.validate_not_after()?;
+        Ok(())
+    }
+}
+
+
+//------------ SubjectPublicKeyInfo ------------------------------------------
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct SubjectPublicKeyInfo {
+    algorithm: PublicKeyAlgorithm,
+    subject_public_key: BitString,
+}
+
+impl SubjectPublicKeyInfo {
+    pub fn decode<S: Source>(source: S) -> Result<Self, S::Err> {
+        Mode::Der.decode(source, Self::take_from)
+    }
+
+    pub fn take_from<S: Source>(
+        cons: &mut Constructed<S>
+    ) -> Result<Self, S::Err> {
+        cons.take_sequence(|cons| {
+            Ok(SubjectPublicKeyInfo {
+                algorithm: PublicKeyAlgorithm::take_from(cons)?,
+                subject_public_key: BitString::take_from(cons)?
+            })
+        })
+    }
+
+    pub fn key_identifier(&self) -> Digest {
+        digest::digest(
+            &digest::SHA1,
+            self.subject_public_key.octet_slice().unwrap()
+        )
+    }
+}
+
+
+//------------ PublicKeyAlgorithm --------------------------------------------
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum PublicKeyAlgorithm {
```

Comment:
Side note: I want to move all these algorithm related things into a module of their own and let them provide the actual crypto primitives, too, so we can more easily add algorithms later.